TFLu - refactor apollo3

oclyke opened the pull request:
first step in revamping ambiq apollo3 support in TFLu as discussed in #44737

this PR, applied in isolation, breaks examples. it requires the three accompanying PRs listed below to fix the corresponding examples:
* person detection #44954
* magic wand #44955 
* micro speech #44957

Commit:
remove redundant apollo3 implementations

Commit:
add generic apollo3 implementation

Commit:
update test_all.sh to use test_apollo3.sh

Commit:
allow GENDIR to handle unique BOARDs

Commit:
Update .gitignore

advaitjain commented on tensorflow/lite/micro/tools/make/targets/apollo3_makefile.inc:
```diff
@@ -67,13 +75,13 @@ LDFLAGS += \
-ifeq ($(TARGET), apollo3evb)
+ifeq ($(BOARD), apollo3evb)
   BOARD_BSP_PATH := $(APOLLO3_SDK)/boards/apollo3_evb/bsp
 endif
-ifeq ($(findstring sparkfun,$(TARGET)), sparkfun)
-  BOARD_BSP_PATH := $(APOLLO3_SDK)/$(SF_BSPS_DEST)/$(subst sparkfun_,,$(TARGET))/bsp
+ifeq ($(findstring sparkfun,$(BOARD)), sparkfun)
+  BOARD_BSP_PATH := $(APOLLO3_SDK)/$(SF_BSPS_DEST)/$(subst sparkfun_,,$(BOARD))/bsp
```
let's make this an if-else block. something like

```make
ifeq ($(BOARD), apollo3evb)
else ifeq ($(BOARD), sparkfun_edge)
else
  $(error "Unsiupported board: $(BOARD)")
ebdif
```

advaitjain commented on tensorflow/lite/micro/tools/make/targets/apollo3_makefile.inc:
```diff
@@ -67,13 +75,13 @@ LDFLAGS += \
   -nostdlib --specs=nano.specs -t -lstdc++ -lc -lnosys -lm \
   -Wl,-T,$(TENSORFLOW_ROOT)$(APOLLO3_SDK)/boards/apollo3_evb/examples/hello_world/gcc_patched/apollo3evb.ld \
-  -Wl,-Map=$(TENSORFLOW_ROOT)$(MAKEFILE_DIR)/gen/$(TARGET).map,--cref
+  -Wl,-Map=$(TENSORFLOW_ROOT)$(MAKEFILE_DIR)/gen/$(BOARD)_$(TARGET).map,--cref
```
if an incorrect board were to be specified, it would be nicer to have a clear error message about that, rather than some linker error about file not found.

I might suggest moving the if-else block for the board below to the top of the file so that you can provide an incorrect board error before anything else.

advaitjain commented on tensorflow/lite/micro/tools/make/targets/apollo3_makefile.inc:
```diff
@@ -67,13 +75,13 @@ LDFLAGS += \
   -Wl,-T,$(TENSORFLOW_ROOT)$(APOLLO3_SDK)/boards/apollo3_evb/examples/hello_world/gcc_patched/apollo3evb.ld \
-  -Wl,-Map=$(TENSORFLOW_ROOT)$(MAKEFILE_DIR)/gen/$(TARGET).map,--cref
+  -Wl,-Map=$(TENSORFLOW_ROOT)$(MAKEFILE_DIR)/gen/$(BOARD)_$(TARGET).map,--cref
 BUILD_TYPE := micro
```
BUILD_TYPE:=micro is unnecessary and can be removed.

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_apollo3.sh:
```diff
@@ -0,0 +1,48 @@
+
+# source tensorflow/lite/micro/tools/ci_build/helper_functions.sh
+
+# define a method to add apollo3 targets
+TARGETS=''
+add_target () {
+  TARGETS+=$1
+  TARGETS+=' '
+}
+
+# add all apollo3 targets
+add_target apollo3evb
+add_target sparkfun_edge
```
also, let's start by only having ```TARGETS=sparkfun_edge``` (so that the behavior matches what we currently have).

In general, I want to make sure that the CI does not take too much time (since it is run as part of *every* pull request on tensorflow).

We can add apollo3evb to the CI as a separate PR. if we feel like the additional time is justified. We can also investigate having a nightly build with a build badge instead.

advaitjain reviewed (changes requested):
Thanks for the PR and taking on this improvement and refactor!

In addition to my review comments,  a few more items before we can merge this change:

 * Add to your PR description exactly what tests you have run. At minimum, I would say that you should have built and run the examples (hello_world, magic_wand, micro_speech, person_detection, person_detection_experimental)  and confirmed that they work (my guess is that there might be a few small tweaks needed to have them work).

  * Update the readme's of the examples to add the new make command

 * Have a `micro/tools/make/targets/sparkfun_edge_makefile.inc` that prints out an error message and gives the new command to run. This will make it easier for the existing users to figure out what the new invocation is.

 * If you aren't already a member of the [SIG-micro](https://groups.google.com/a/tensorflow.org/g/micro) mailing list, it would be great if you could join and email that we're in the process of changing how TFLM is built for sparkfun edge (and point to this PR and the bug)

After this PR is merged:
 * Email SIG-micro again to let people know that the change is merged.

Commit:
add error with redirect message

when users try TARGET=sparkfun_edge they will be notified of the proper command structure

advaitjain commented:
applied the awaiing response label so that the @gbaned role doesn't request review from me.

@oclyke feel free to let me know when this is ready for review again. No rush, thanks!

tensorflowbutler commented:
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

gbaned commented:
@oclyke  Any update on this PR? Please. Thanks!

tensorflowbutler commented:
It has been 29 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
